ironfish-zkp/src/circuits/util.rs

```diff
@@ -143,3 +144,35 @@ pub fn assert_valid_asset_generator<CS: bellperson::ConstraintSystem<blstrs::Sca
 
     Ok(())
 }
+
+pub trait Reader: Sized {
+    fn read<R: std::io::Read>(reader: &mut R) -> Result<Self, std::io::Error>;
```

Comment:
`reader: &mut R` -> `mut reader: R`

Reply:
Why is this way preferable? I was reading that the way you suggest takes ownership, whereas the current way is borrowed, is it better in the case of readers to take ownership?

Reply:
`Read` is already automatically implemented for `&mut R`: https://doc.rust-lang.org/std/io/trait.Read.html#impl-Read-for-%26mut+R. This means that if you have a method with signature `fn foo<R: Read>(reader: R) -> ...`, both `foo(reader)` and `foo(&mut reader)` are accepted. Conversely, if you have `fn foo<R: Read>(reader: &mut R) -> ...`, only `foo(&mut reader)` is accepted. Therefore, accepting `R` allows opening the door to more types.

In the particular case of `Read` (or `Write`), this is especially useful because readers (or writers) may be wrappers that contain references to other structs, so requiring a `&mut R` often results in a reference to a reference, and this effect is amplified when adding layers of wrappers.